_episodes/12.md

```diff
@@ -0,0 +1,263 @@
+
+
+## Concept of the week: Trino client, Python, and Apache Superset 
+
+What is the general data flow from a connected data source?
+
+ * Trino workers request data from the data source with specific connector
+ * Workers process data and send it to the coordinator
+ * Coordinator does final processing
+ * Supplies the data via HTTP / REST stream to requestor
+ * Requestor is a “client” such as JDBC driver, or Trino CLI
+ * Client translates data further and provides to application (Java application
+   using JDBC driver) or user interface/directly to user (output in CLI)
+ * User views part of data and scrolls down
+ * Client requests more data from coordinator via HTTP / REST (and see above)
+
+What clients are provided by Trino project?
+
+ * JDBC driver
+ * Trino CLI
+ * Trino Python client
+ * Trino Go client
+
+What other clients are there?
+
+ * ODBC driver from Starburst
+ * Various other clients from the open source community 
+  * R
+  * NodeJS/Javascript
+ * See https://trino.io/resources.html#libraries
+
+What happens in the Python world?
+
+Disclaimer: I am not a Pythonista or Pythoneer.
+
+ * DB-API 2.0 
+   * PEP 249 https://www.python.org/dev/peps/pep-0249/
+   * Python standard library
+ * trino-python-client 
+   * Wraps complexity of Trino HTTP / REST
+   * Supports authentication and such
+   * Provides DB API endpoints / implementation
+   * Preferred method to query Trino
+ * SQLAlchemy https://www.sqlalchemy.org/
+   * SQL toolkit
+   * ORM mapper
+   * Widely used, eg. in Apache Superset
+   * Supports dialects
+ * PyHive
+   * Not really a SQL wrapper
+   * Aimed at Hive QL
+   * Only kind of useful for Trino, limited compatibility
+ * JDBC driver (Java !) and PySpark
+   * Possible, but a hack really
+ * PyJDBC 
+   * Wraps DB API around any JDBC driver
+   * Kind of a hack since it goes through JDBC to HTTP, when Trino python
+   client does the same more directly
+ * PyODBC 
+   * Similar hack to PyJDBC
+ * Potentially also possible to talk to via HTTP directly
+   * That’s like reimplementing the trino-python-client
+   * Also see question of the week later
+
+Beyond that, it will vary from application to application.
+
+Let’s find out from our guests how this hangs together in Apache Superset, since
+it is using Python.
+
+## PR of the week: Superset PR 13105 feat: first step native support Trino
+
+In this week's pull request <https://github.com/apache/superset/pull/13105> that
+was graciously added by [dungdm93](https://github.com/dungdm93).
+
+The first thing we need to understand about this addition is the concept of a
+database engine in Superset. A database engine handles a lot of the custom
+interactions between various databases and maps them to the interface that 
+Superset understands. If certain concepts are missing in a certain database, 
+like time granularity or SQL syntax, the database engine for that database
+will indicate to Superset that this is not available, and it won't show up as an
```

Comment:
that database indicated to Superset that this is not available. As a result the option does not show in Superset or a concise error message is reported.